Make `this` variable available in template entities

akloeckner opened the pull request:
## Proposed change
This PR makes the variable `this` available in template entities, which simplifies the use of self-referencing template entities.
Because, without this, we have to repeat the entity id every time.
If we can solve this without explicitly spelling the entity id, code can be re-used much better.

As a side-effect, this will allow to use `variables`-like patterns,
where attributes can be used as variables to calculate subsequent attributes or state.

Example:
```yaml
template:
  sensor:
    - name: test
      state: "{{ this.attributes.test }}"
      # not: "{{ state_attr('sensor.test', 'test' }}"
      attributes:
        test: "{{ now() }}"
```



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/22211

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

akloeckner committed:
feat: make this variable available in template entities

This makes the variable `this` available in template entities.
It will simplify the use of self-referencing template entities.
Because, without this, we have to repeat the entity id every time.
If we can solve this without explicitly spelling the entity id,
code can be re-used much better.

As a side-effect, this will allow to use `variables`-like patterns,
where attributes can be used as variables to calculate subsequent attributes or state.

Example:
```yaml
template:
  sensor:
    - name: test
      state: "{{ this.attributes.test }}"
      # not: "{{ state_attr('sensor.test', 'test' }}"
      attributes:
        test: "{{ now() }}"
```

emontnemery commented:
We should perhaps align this with the recent change in MQTT which adds the MQTT entity's `entity_id` as a variable to all MQTT templates: https://github.com/home-assistant/core/pull/62105

akloeckner commented:
> We should perhaps align this with the recent change in MQTT which adds the MQTT entity's `entity_id` as a variable to all MQTT templates: #62105

While this is quite an extensive change which I don't understand in its entirety, I agree that the variables available to a template should be the same all the time. So, basically, `template`, `mqtt` and also `automation` should all follow the same rule. Now that we seem to have two rules, we should think about what is the better one...

| `this` | `entity_id` |
|---|---|
| * directly exposes the whole state object | * forces the user to use `states()` and its friends   |
| * can be abused in strange ways | * is more complex to use |
| * might be hard to track for changes | * should track changes just fine through standard interfaces |

I think you're right: `entity_id` is the robuster choice despite its higher complexity to use. Not sure, what that means for `this` in `automation`...

akloeckner committed:
expose entity_id instead of this

akloeckner commented:
I exposed `entity_id` now, instead of `this`. The new minimal example would be:
```
template:
  sensor:
    - name: test
      state: "{{ state_attr(entity_id, 'test') }}"
      attributes:
        test: "{{ now() }}"
```

akloeckner committed:
add test

frenck commented:
I honestly think that "this" is a better option to go for, for now and the future.

Especially since there is another big difference between entity_id (and next using methods) and "this".

"This" contains the full state at the moment of triggering.

akloeckner commented:
> at the moment of triggering.

<details>
  <summary>I verified using this (collapsed) example.</summary>

```yaml
input_text:
  text1:
    name: Text1
    initial: Nothing
  text2:
    name: Text2
    initial: Nothing
automation:
  alias: "Test"
  id: test
  mode: parallel
  trigger:
    - platform: homeassistant
      event: start
  action:
    - service: input_text.set_value
      target:
        entity_id: input_text.text1
      data:
        value: "{{ this.attributes.last_triggered }}"
    - delay: 5
    - service: input_text.set_value
      target:
        entity_id: input_text.text2
      data:
        value: "{{ this.attributes.last_triggered }}"
    - service: automation.turn_off
      target:
        entity_id: automation.test
```

This results in two different `input_text`s, if you trigger the automation twice within the 5s period.
</details>

I updated the table above accordingly. However, I have the impression, that this behavior will be more difficult to implement, because the variables for the templates are set in one place (which I changed in this PR) during setup, but `this` would need to be freezed for selected components only ("triggered" sensors, components with `delay`'able script parts). In my first commit, I verified the example to work, so `this` did actually change after setup. Additionally, `this` might need to be set up to trigger a template re-render.

I'm not sure how to proceed. Although I am no friend of compromises in these matters: Maybe the way to go is to provide both variables where-ever possible. So, start with `entity_id` here, because it is easy. And add `this` with more thought, later.

frenck commented:
> However, I have the impression, that this behavior will be more difficult to implement, because the variables for the templates are set in one place

I don't see how that would more or less difficult to implement?

> So, start with entity_id here, because it is easy. And add this with more thought, later.

I don't agree actually. With `this`, the `entity_id` would also be available (as `this.entity_id`), so setting sail to implementing both seems odd.

akloeckner commented:
> I don't see how that would more or less difficult to implement?

Because we have to distinguish between "instantaneous" templates, such as sensors, and "`delay`'able" templates, such as lights:
   * For the former, we can just pass the state object and it will update. That is easily done in the same place as `entity_id` is now.
   * For the latter, we have to somehow "freeze" the object once the scripts are started. For that I hve no clue currently.

> I don't agree actually. With `this`, the `entity_id` would also be available (as `this.entity_id`), so setting sail to implementing both seems odd.

I agree: I am no friend of compromises in these matters. So, let's take the time to decide thoughtfully. :-) I'll update the table above with the remark about `this.entity_id`.

frenck commented:
> For the former, we can just pass the state object and it will update

We should not pass in the state object as it, it should always be a deep copy from the moment it triggered. This is in line with automations and scripts. We should not invent new constructs here.

frenck commented:
> And: We will have to change the complete architecture of template entities. Because variables are passed to the TrackTemplate only during setup. At least as far as I can see. Maybe I don't see far enough...

We can always add it last minute before rendering the template (if not added yet).

akloeckner commented:
> We can always add it last minute before rendering the template (if not added yet).

Can we? I think, this is true for `TriggerEntities` here: https://github.com/home-assistant/core/blob/1aceacb91836a4438bd42f8761dcfc891f9e81cd/homeassistant/components/template/trigger_entity.py#L127

But for `TemplateEntities`, a template result is tracked and there is no way to intervene from within `TemplateEntity`: https://github.com/home-assistant/core/blob/1aceacb91836a4438bd42f8761dcfc891f9e81cd/homeassistant/components/template/template_entity.py#L385

Anyway:
1. Let's first decide on `this` vs. `entity_id`. 
2. Let's then decide whether `this` should be frozen at trigger time (whereever that has a meaning, e.g. `TriggerEntity`). And whether it is allowed to change after startup for the other cases (e.g. `TemplateEntity`).
3. And then, let's think about how to implement.

For the first two points, I feel you (@frenck and maybe @emontnemery) have stronger opinions (and background) than I have. I would still go for `entity_id`. Mainly because it's cleaner and because I couldn't come up with a use case for the only additional "feature" of `this`, i.e., freezing `this` at trigger time.

akloeckner commented:
Since my last post is quite some time ago: any idea how to move on with this? I'd really like to simplify my many template sensors. But I'd also like to avoid searching for ways to implement "frozen" `this`, especially as long as it's not 100% settled which is the best solution from the user's and structure perspective.

emontnemery commented:
I think allowing accessing the entity's state via a `this` variable as @frenck suggests would be really nice.

I think the `this` variable could be very similar to a `TemplateState`, but instead of passing in a `State` when instantiating, we pass in the `entity_id`.

Additional tests are needed to ensure accessing the `this` variables triggers re-rendering correctly, and not accessing the `this` variable does not result in re-rendering.

emontnemery committed:
Refactor to expose this variable

emontnemery committed:
Tweak repr dunder

frenck reviewed (approved):
Super nice! ❤️ 

Needs documentation though!

akloeckner commented:
Thanks @emontnemery, I also like your solution!

I meant to install and run your changes on my development system to check them and add some things:
* Understand, why the new `TemplateState` variant is needed.
* Add test for changing `this` not triggering re-render, if not used.
* Add `this` to `TriggerEntities`, too.

But I failed. I might have been too tired and will retry soon again...

emontnemery commented:
@akloeckner since I've been pushing to your branch, the easiest way may be to do something like this if you also have local changes (ignore this if you didn't have problems with syncing the commits):
```
git branch -m feat/template-this feat/template-this_old
git fetch origin
git checkout feat/template-this_old
```

Now your local changes are on branch `feat/template-this_old`, and you have a new local branch `feat/template-this_old` tracking the PR branch.

emontnemery commented:
BTW, your suggested actions (new tests, add this also to trigger-based templated entities etc. are all good), please push new tests to this PR. Adding support to other entities can be done in follow-up PRs I think.

Can you confirm what you mean by this:
> Maybe, the feature should directly be added to more types of entities. At least the use case of setting an icon will be valid for 
> any template entity. And it will be confusing, if this is only available in selected types.

Do you mean making the `this` variable available also in other integrations which support templating attributes?

akloeckner commented:
> ignore this if you didn't have problems with syncing the commits

I had the development set up in a docker container, but somehow destroyed that. Now, I am struggling to just get it back to work. :-/

> please push new tests to this PR

I'll do that, as soon as I get up and running again...

> Do you mean making the this variable available also in other integrations which support templating attributes?

I read the template integration docs to see where I would have to add the new feature. And they are quite involved and confusing already. If now, `this` is only available in "state-based" (I think is the term, as opposed to trigger-based) "sensor" entities, that will be even more confusing. It should at least be available in *all* sensor entities (of the template integration). Even better, if it were available in all entities set up in the template integration.

So, no: I did not mean other integrations. But the template integration itself has quite a lot of entity types available...

emontnemery commented:
With this PR (both your original version, and after changing to `this`) the new variable should be exposed to templates in all state-based template entities as they all inherit `TemplateEntity`, I think? 
We should add tests to make sure that's really the case though, maybe a test in test_init.py, parametrized for all entity types.

What's missing are the trigger-based template entities.

I'll update the docs PR too so all entities are covered.

akloeckner commented:
> the new variable should be exposed to templates in all state-based template entities as they all inherit TemplateEntity

I begin to see, why Franck called your change "super nice"...

I'll make sure to cover that in tests. I hope, I will (given my setup is down)...

emontnemery commented:
Let's merge this as is, additional tests can be added in follow-up PRs 👍

akloeckner commented:
👍 thanks. I'm still not set up. So, probably the best way to proceed...
